src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -706,7 +742,20 @@ public byte[] Http2AltSvcOriginUri
                     Trace("Attempting new HTTP3 connection.");
                 }
 
-                QuicConnection quicConnection = await ConnectHelper.ConnectQuicAsync(authority.IdnHost, authority.Port, _sslOptionsHttp3, cancellationToken).ConfigureAwait(false);
+                QuicConnection quicConnection;
+                try
+                {
+                    quicConnection = await ConnectHelper.ConnectQuicAsync(authority.IdnHost, authority.Port, _sslOptionsHttp3, cancellationToken).ConfigureAwait(false);
+                }
+                catch
+                {
+                    // Disables HTTP/3 until server announces it can handle it via Alt-Svc.
+                    lock (SyncObj)
+                    {
+                        ExpireAltSvcAuthority();
```

Comment:
Oh, this should be done without a lock.

Reply:
I'm interested in reasoning for doing it outside of lock, could you elaborate on this?

Reply:
> Apparently we renamed "blacklist" to "blocklist" and didn't account for when "blacklist" is used as a verb. Now this method name is bad. If you want to clean that up, I'd rename as AddToAltSvcBlockList or something like that.

I'm not following. Firstly I'm not touching anything with `blocklist` in its name, I'm not even using such methods. Secondly, `ExpireAltSvcAuthority` only nulls `_http3Authority` nothing else. It doesn't add it to blocklist: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L936-L940

Reply:
Sorry, there are a few points. To clarify:

An accepted pattern is to *always* send Alt-Svc on every response. So if we only remove the current authority, this will cause us to ping-pong between HTTP/1 and HTTP/3 over and over again. Adding the service to the blocklist, which will prevent the Alt-Svc authority from being used for a small period of time, will prevent this ping-pong behavior.

`ExpireAltSvcAuthority` *does* need to be done inside a lock, but `BlocklistAuthority` does not -- it acquires its own locks.

The method name when I wrote it was "BlacklistAuthority", but a recent inclusiveness PR renamed "Blacklist" to "Blocklist". I was just noting that "Blocklist" is not a verb, so that PR gave this method a nonsensical name. If you wanted to fix that here you could rename the `BlocklistAuthority` method to something like `AddAltSvcAuthorityToBlockList`.

Reply:
I see, now I understand, I will change it to call `BlocklistAuthority` (outside of lock).
I won't change the naming though. That's probably an issue on its own and separated from this PR.

Reply:
According to [`BlocklistAuthority`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1012) we shouldn't use it for `_originAuthority`.
But in our case it could be just that. For H3 and >= Version we assume prenegotiated H3 and use `_originAuthority`.